azkaban-db/src/main/java/azkaban/db/DatabaseOperator.java

```diff
@@ -117,16 +118,35 @@ public DatabaseOperator(final QueryRunner queryRunner) {
    * @return The number of rows updated.
    */
   public int update(final String updateClause, final Object... params) throws SQLException {
-    try {
-      return this.queryRunner.update(updateClause, params);
-    } catch (final SQLException ex) {
-      // todo kunkun-tang: Retry logics should be implemented here.
-      logger.error("update failed", ex);
-      if (this.dbMetrics != null) {
-        this.dbMetrics.markDBFailUpdate();
+    int retry = 0;
+    SQLException exception;
+    String errorMsg = "Update failed: Reached maximum number of retries: " + AzDBUtil.MAX_RETRIES_ON_DEADLOCK;
+    do {
+      try {
+        return this.queryRunner.update(updateClause, params);
+      } catch (final SQLException ex) {
+        exception = ex;
+        if (ex.getErrorCode() == 1213) { // 1213 -> Deadlock error
+          retry++;
+          logger.warn("Deadlock detected when trying to execute: " + updateClause + " with values: "
+              + Arrays.toString(params));
+          try {
+            Thread.sleep(AzDBUtil.RETRY_WAIT_TIME);
```

Comment:
Using exponential backoff while sleeping is preferred, something like Thread.sleep(AzDBUtil.RETRY_WAIT_TIME * retry_count)

Reply:
Update flow query is called many times during a flow execution and it's a synchronized method. By adding exponential backoff here we're adding delays to the execution time that collectively could go into minutes.